src/Orchard.Web/Modules/Orchard.Setup/Services/SetupService.cs

```diff
@@ -194,19 +195,39 @@ await scope
                 }
             }
 
-            shellSettings.State = TenantState.Running;
-            _orchardHost.UpdateShellSettings(shellSettings);
             return executionId;
         }
 
-        private async Task<string> CreateTenantDataAsync(IServiceScope scope, SetupContext context, ShellContext shellContext)
+        private async Task<string> CreateTenantDataAsync(IServiceScope scope, 
+            SetupContext context, 
+            ShellContext shellContext)
         {
             var recipeManager = scope.ServiceProvider.GetService<IRecipeManager>();
 
             var recipe = context.Recipe;
             var executionId = await recipeManager.ExecuteAsync(recipe);
 
+            // Once the recipe has finished executing, we need to update the shell state to "Running", so add a recipe step that does exactly that.
```

Comment:
Why the choice to use a recipe step instead of calling the logic directly. I assume it's by design that the shell has to be started when the setup is done. And isn't it already doing it out of the box?

And same thing for when a feature is enabled/disabled from a recipe, it should by itself restart the shell, using a deferred task probably. But I like the idea of having a step though, one might want to force it at some point.